RFCXXX-PSRemoting-Providers.md

```diff
@@ -0,0 +1,298 @@
+
+Custom provider are even more complicated where the user must manage the session lifetime themselves by using a cmdlet provided by the module that implemented the provider.
+
+```powershell
+$session = New-CustomTransportSession -Target ...
+try {
+    Invoke-Command -Session $session { 'test' }
+}
+finally {
+   $session | Remove-PSSession
+}
+```
+
+This has now turned a single cmdlet call into something more complex and error prone with the introduction of the `try/finally` that users could omit from their code.
+As well as this, each provider needs to implement their own `New-CustomTransportSession` cmdlet rather than utilitise the hard work already done in `Invoke-Command`, `New-PSSession`, `Enter-PSSession`, etc.
+This can lead to substandard error messages, cancellation problems, as the current interface for doing so has a few complexities around it.
+It's also complicated further as now the creation of the session must be done by the user and the fan-out feature of `Invoke-Command` is no longer available.
+For example the following can be used to connect to 2 hosts in parallel using the builtin SSH provider:
+
+```powershell
+Invoke-Command -HostName Host1, Host2 -ScriptBlock { 'test' }
+```
+
+Trying to implement this for custom providers would require some complicated solution like the following:
+
+```powershell
+$sessions = 'Host1', 'Host2' | ForEach-Object -Parallel {
+    New-CustomTransportSession -Target $_
+}
+
+try {
+    Invoke-Command -Session $session { 'test' }
+}
+finally {
+    $sessions | Remove-PSSession
+}
+```
+
+Introduction the `-Parallel` block itself even brings in more complications, for example `$using:var` is required when referring to vars outside of the scriptblock.
+What was a single cmdlet has now turned into multiple lines without any real error handling.
+This can be a fairly large obstacle for custom transport providers to overcome when it comes to getting people willing to try out their code vs what's built into PowerShell itself.
+
+## Proposed Solution
+
+The proposed solution would be to implement a registry of remoting providers that modules can register to.
+This registration would include a unique prefix allowing users to simply specify the transport type as part of the host connection string.
+More examples can be seen in the `User Experience` header but in short it would turn this:
+
+```powershell
+$sessions = 'Host1', 'Host2' | ForEach-Object -Parallel {
+    New-CustomTransportSession -Target $_
+}
+
+try {
+    Invoke-Command -Session $session { 'test' }
+}
+finally {
+    $sessions | Remove-PSSession
+}
+```
+
+Into
+
+```powershell
+Invoke-Command -ComputerName custom:Host1, custom:Host2
+```
+
+Existing providers can also have their own host prefix allowing things like `-ComputerName ssh:Host1` to connect over SSH to `Host1`.
```

Comment:
Would it make sense to just adopt URL syntax with the scheme being the transport?

Reply:
It certainly is possible and wouldn't hurt. I didn't have it as part of the initial proposal because it is 2 extra chars (`//`) and I didn't particularly want to define the structure of the string, rather I wanted it to be up to the provider itself as to what format it expects of the value.

Reply:
+1 on the URL format, it's just a few characters away and it would be a lot more future-proof

Reply:
To me what are the benefits of using the URL format? My concerns is tying the provider to URL only when a provider might want something less restrictive. I don't think this value should be too complex and providers wanting to support custom connection options should be encouraged to use the `-SessionOption` parameter instead.